chalk-solve/src/clauses/syntactic_eq.rs

```diff
@@ -0,0 +1,165 @@
+use std::{iter, mem::replace};
+
+use chalk_engine::fallible::Fallible;
+use chalk_ir::{
+    cast::Cast,
+    fold::{shift::Shift, Fold, Folder, SuperFold},
+    interner::Interner,
+    AliasEq, AliasTy, Binders, BoundVar, DebruijnIndex, Goal, GoalData, Goals, ParameterKind,
+    ParameterKinds, ProgramClause, ProgramClauseData, ProgramClauseImplication, QuantifierKind, Ty,
+    TyData,
+};
+
+pub fn syn_eq_lower<I: Interner, T: Fold<I>>(interner: &I, clause: &T) -> <T as Fold<I>>::Result {
+    let mut folder = SynEqFolder {
+        interner,
+        new_params: vec![],
+        new_goals: vec![],
+        binders_len: 0,
+    };
+
+    clause
+        .fold_with(&mut folder, DebruijnIndex::INNERMOST)
+        .unwrap()
+}
+
+struct SynEqFolder<'i, I: Interner> {
+    interner: &'i I,
+    new_params: Vec<ParameterKind<()>>,
+    new_goals: Vec<Goal<I>>,
+    binders_len: usize,
```

Comment:
Stores the number of variables in the binder that we are adding parameters into thus far. Incremented for each new variable we add.